consensus/src/common.rs

```diff
@@ -103,88 +77,93 @@ impl Chain<ProtocolStateLegacy> for ProtocolStateChain {
+    /// The function the length of a chain. The output is the length of the chain in blocks.
+    pub fn length(&self) -> usize {
         self.consensus_state()
             .map(|s| s.blockchain_length.0 as usize)
             .unwrap_or(0)
     }
 
-    fn last_vrf_hash_digest(&self) -> Result<String, ConsensusError> {
+    /// This function returns the hex digest of the hash of the last VRF output
+    ///  of a given chain. The input is a chain C and the output is the hash digest.
+    pub fn last_vrf_hash_digest(&self) -> Result<String, ConsensusError> {
         let hash = self
-            .top()
+            .consensus_state()
             .ok_or(ConsensusError::TopBlockNotFound)?
-            .body
-            .consensus_state
             .last_vrf_output
             .digest();
         Ok(hex::encode(hash))
     }
 
-    fn state_hash(&self) -> Option<Fp> {
+    /// This function returns hash of the top block's protocol state for a given chain.
+    /// The input is a chain C and the output is the hash.
+    pub fn state_hash(&self) -> Option<Fp> {
         self.top().map(|s| s.state_hash_fp())
     }
 }
 
-/// A trait that defines operations for consensus calculation
-pub trait Consensus {
-    /// Chain type
-    type Chain;
+/// A trait that defines operations for chain selection
+pub trait ChainSelection {
     /// Top level API to select between chains during a fork.
-    fn select_secure_chain<'a>(
-        &'a self,
-        candidates: &'a [Self::Chain],
-    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+    fn select_secure_chain(&mut self, candidates: Vec<Self>) -> Result<(), ConsensusError>
+    where
+        Self: Sized;
 
     /// Selects the longer chain when there's a short range fork.
-    fn select_longer_chain<'a>(
-        &'a self,
-        candidate: &'a ProtocolStateChain,
-    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+    fn select_longer_chain(&mut self, candidate: Self) -> Result<(), ConsensusError>
+    where
+        Self: Sized;
 
     /// Checks whether the fork is short range wrt to candidate chain
-    fn is_short_range(&self, candidate: &ProtocolStateChain) -> Result<bool, ConsensusError>;
+    fn is_short_range(&self, candidate: &Self) -> Result<bool, ConsensusError>;
 
     /// Calculates the relate minimum window density wrt to candidate chain.
-    fn relative_min_window_density(
-        &self,
-        candidate: &ProtocolStateChain,
-    ) -> Result<u32, ConsensusError>;
+    fn relative_min_window_density(&self, candidate: &Self) -> Result<u32, ConsensusError>;
 
     /// Constants used for consensus
     fn config(&self) -> ConsensusConstants;
 }
 
-impl Consensus for ProtocolStateChain {
-    type Chain = ProtocolStateChain;
-    fn select_secure_chain<'a>(
-        &'a self,
-        candidates: &'a [Self::Chain],
-    ) -> Result<&'a ProtocolStateChain, ConsensusError> {
-        let tip = candidates.iter().fold(Ok(self), |tip, candidate| {
```

Comment:
Looks like there was a bug, with `fold` fn, all `self` usage should be replaced by `tip`